Add support for GitLab, Gitee, and SourceForge

csandman opened the pull request:
As the title says, this PR adds support for GitLab, Gitee, and SourceForge, and would close #2.  GitLab was originally requested in that issue, but the other two were not. Like I had mentioned in a comment, I've looked through a bunch of these extensions, and some of them have support for Gitee and some have support for SourceForge, so I decided to try adding both of them.  The results look pretty good IMO.

Before I talk about those, I made two other small general tweaks that I'd like to point out.

1. I switched the filename getter to take the first part of a path (split on `/`) if the name is representative of a path.  For example, when there's only one folder inside another folder, GitHub concatenated the names and clicking sends you straight through to the second folder.  In general, I've found that the top level folder in these situations is more likely to have a custom icon, so I used the first part of the path.  However, it could technically make more sense to use the end part of the path, as that's where the link goes.  Either way, up to you whether you want to keep this change.

<img width="919" alt="image" src="https://user-images.githubusercontent.com/9214195/176955549-cbcbdc0a-b8c0-481d-9c33-241388f58388.png">

2. In order to get these new providers working, I changed the structure of the `getIsDirectory`, `getIsSubmodule`, and `getIsSymlink` functions to accept an object with the shape `{ row, icon }`.  In some cases, you can not determine these features from the icon itself, and having access to the `row` element gives a lot more freedom for figuring that out. It was possible to get the same features from the row by chaining `.parentNode` but I've never felt super comfortable with traversing up a DOM tree.  Plus, it's still about as simple to implement these functions as you can just destructure the part you need:

```js
getIsDirectory: ({ icon }) => icon.getAttribute('data-testid') === 'folder-icon',
getIsSubmodule: ({ row }) => row.querySelector('a')?.classList.contains('is-submodule') || false,
```

So here's some info on the new providers:

## GitLab

Like I mentioned, the main purpose of adding GitLab support is so that the folders get icons.  Besides that it is mostly similar.

<img width="1101" alt="image" src="https://user-images.githubusercontent.com/9214195/176956332-9a8e525a-c906-4c5b-a912-9f997bcd5d0d.png">

One extra feature I added however, was replacing the icon for the Readme shown below the list of files:

<img width="1261" alt="image" src="https://user-images.githubusercontent.com/9214195/176956416-ecce48c6-08a7-4897-b9c4-e4b2f2c25cbd.png">

As well as the icon you see when viewing a specific file:

<img width="1447" alt="image" src="https://user-images.githubusercontent.com/9214195/176956520-ce4be3db-078f-4578-b7a9-8dff35d60870.png">

I'm sure you have plenty of repos you can test this with as you mentioned you're now using GitLab for work, but here's an instance of that test repo I shared with you that I pushed to GitLab: https://gitlab.com/csandman/test

## Gitee

I'm not gonna lie, I had never heard of Gitee until I saw another extension supporting it, but it has a very similar layout to GitLab.  I did the same replacements as I did on GitLab, the file table, readme title, and individual file titles:

<img width="1334" alt="image" src="https://user-images.githubusercontent.com/9214195/176956850-db939f8a-8392-4d9c-8d8d-fb8eaed0b040.png">

<img width="1330" alt="image" src="https://user-images.githubusercontent.com/9214195/176956922-156d6567-94a8-4a0b-8371-6a870a89cc59.png">

I tested the symlink/submodule getters using that same repo but I don't feel like making it public because they have an approval process for that and I don't want to deal with it.  You can test it with any of the repos on their explore page: https://gitee.com/explore

## SourceForge

For SourceForge, I had to handle things a little differently than the rest.  Each project has both a code page as well as a files page with a list of downloadable output files.  I decided both would benefit from this replacement.  However, on the Files page, only the folders have an icon by default, so I targeted the element wrapping the icon, replaced the child icon if it does exist, and prepended the new icon if it does not.

Here's the Files page:

<img width="1421" alt="image" src="https://user-images.githubusercontent.com/9214195/176957337-1cbadffb-777a-448e-b14a-8767034b0919.png">

And here's the Code page:

<img width="1467" alt="image" src="https://user-images.githubusercontent.com/9214195/176957383-388635b3-2194-41a2-947b-fc95fc983cb8.png">

You can check out the code page for my test repo mentioned above here: https://sourceforge.net/p/csandman-test/code/ci/main/tree/

And you can check the files page for any of the public projects listed here: https://sourceforge.net/directory/

Let me know if anything needs changing!

Claudiohbsantos commented on src/providers/gitlab.js:
```diff
@@ -0,0 +1,28 @@
+    // Cell in file list, file view header, readme header
+    filename: 'td.tree-item-file-name, .file-title-name, .gl-link',
+    // Any icon not contained in a button
+    icon: 'svg:not(.gl-button-icon)',
```
I'd rather keep the selectors contained to only match within what we identified as a row. This way if gitlab decides to add svgs elsewhere in the page we are less likely to break.

Maybe a simple[ descendant selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Descendant_combinator) might do the trick here?
```suggestion
    icon: '.tree-item svg',
```

csandman replied:
For this one, I could not find a more specific selector that would be able to target the icon in the file headers (both viewing a file, and the readme on the repo's homepage).  The only way I can think to specify it more would be to remove the icons from those other views.

csandman replied:
To specify, the reason for doing `:not(.gl-button-icon)` is to prevent matching the copy button icon in the file headers btw.

Claudiohbsantos commented on src/providers/sourceforge.js:
```diff
@@ -0,0 +1,58 @@
+    // File list row, README header, file view header
+    row: 'table#files_list tr, #content_base tr td:first-child',
+    // File name table cell, file view header
+    filename: 'th, a.icon',
```
Similar note as in gitlab. Can we constrain the selectors for filename/icon a bit more here? I just want to make it as unlikely as possible that we'll match things we don't want to in the long run in order to reduce the maintenance effort needed.

Claudiohbsantos commented on src/providers/sourceforge.js:
```diff
@@ -0,0 +1,58 @@
+    // The iconfont icon not including the delete button icon in the file view header
+    icon: 'a:not(.icon), i.fa',
+  },
+  getIsLightTheme: () => true, // There appears to be no dark theme available for gitee.
```
```suggestion
  getIsLightTheme: () => true, // There appears to be no dark theme available for sourceforge.
```

Claudiohbsantos commented on src/providers/gitlab.js:
```diff
@@ -0,0 +1,28 @@
+  },
+  getIsLightTheme: () => !document.querySelector('body').classList.contains('gl-dark'),
+  getIsDirectory: ({ icon }) => icon.getAttribute('data-testid') === 'folder-icon',
+  getIsSubmodule: ({ row }) => row.querySelector('a')?.classList.contains('is-submodule') || false,
```
Do we need the `|| false` here? If `contains` always returns a boolean we'll always ignore that because `(true || false) === true` and `(false || false) === false`.
```suggestion
  getIsSubmodule: ({ row }) => row.querySelector('a')?.classList.contains('is-submodule'),
```

csandman replied:
The reason for this is that in the case of the file header icons, this element won't get selected (hence the optional chaining), and in that case this would return `undefined`.  And while that wouldn't actually matter, as `undefined` will still count as falsey when the value is checked, my TypeScript brain thought it would be better to keep it consistently returning a boolean haha.  Idk what your thoughts are on that.

Claudiohbsantos replied:
hahah ok, that's fair. Yeah, no particular preference here then. I always just read the result of this method as `contains` or `not contains` but I understand it could be cleaner to enforce a boolean return. lgtm then

Claudiohbsantos commented on src/providers/sourceforge.js:
```diff
@@ -0,0 +1,58 @@
+    }
+    // For the files list, use the anchor element instead of the icon because in some cases there is no icon
+    else {
+      if (iconOrAnchor.querySelector('img')) {
```
Is the idea here that if we find an `img` we know we already replaced the icon and therefore we no-op?

If so I'd recommend adding a more unique attribute to `newSVG` before replacement and checking for that attribute instead. For the same reason as the selector tweaks: just to make it less likely our logic will collide with website changes

Though I don't think I've seen sourceforge change since I was in high school so.... yeah

csandman replied:
Yeah that was the idea!  So I thought having a class could be generally useful when creating a new icon (for other new providers, etc.), so I added the class `material-icons-extension` to the `img` element before passing it into the `replaceIcon` function, and using that to check if the icon has already been added. Let me know if that works!

csandman replied:
Scratch that, I realized this class get's overridden in many of the providers so I moved it inside the SourceForge provider.

Claudiohbsantos reviewed (changes requested):
Thanks for another extremely clean PR! Your descriptions are very easy to follow.

I answered most things in individual comments so I'll skip them here. And I really like the update to the signature of `getIsDirectory` and family. 

Only merge blocking changes are the ones related to the selectors feeling a bit too wide. Let me know if any of them are wide for a particular reason though. Everything else looks great!

csandman commented:
Thanks for checking it out!  I can't remember the exact reasons of the top of my head but some of the selectors are the way they are because of issues with more specific ones not matching properly. I'll check them out and reply to the specific comments with my thought process on Tuesday!

csandman commented on src/main.js:
```diff
@@ -91,7 +100,10 @@ function replaceIcon(itemRow, provider) {
-  const fileName = itemRow.querySelector(provider.selectors.filename)?.innerText.trim();
+  const fileName = itemRow
+    .querySelector(provider.selectors.filename)
+    ?.innerText.split('/')[0] // get first part of path for a proper icon lookup
```
Yeah makes sense, for some reason I thought `innerText` just returned an empty string if there was no content.  Good call!

csandman replied:
And I agree, checking if there is an icon for the child first, and if non exists finding one for the parent probably is a good alternative, but that would involve a bigger change for sure.

csandman commented:
Also, I can't tell if from your comments you're saying you'd like to avoid the icons related to the file headers? If so, that would probably simplify a lot of the selectors, as that's a large part of why they're more general. So let me know if that's the case, I have no problem with removing them!

Claudiohbsantos commented:
Thanks, no rush on this!

And I like the file header icons, no need to remove them. I was mostly thinking of just setting up some basuc hierarchy in the selectors so they only match icons within the tables/rows/container that we expect. We probably don't even need to necessarily change the selectors, just constrain them to only match within certain containers. Not sure if that makes sense, let me know if I'm talking non sense and I can try to explain better

csandman commented on src/providers/gitee.js:
```diff
@@ -0,0 +1,28 @@
+    // File list row, README header, file view header
+    row: '#git-project-content .tree-content .row.tree-item, .file_title, .blob-description',
+    // File name table cell, file view header
+    filename: '[class*="tree-item-"], span.file_name',
```
The main problem with this (and the reason I didn't make it the way you had it) is that it doesn't properly match rows for a submodule.  However, I can add that type of row as a special case if you want, but it might be too specific.  I'll push that and see what you think.

<img width="914" alt="image" src="https://user-images.githubusercontent.com/9214195/177391119-f3de1eb3-0328-4fc8-802d-e97116687d61.png">

csandman commented:
Alright I made the changes you requested, in a way that I hope works for you! I get what you mean now with only matching within the containers the icon/filename are expected to be in.  Like i mentioned in one of my comments, I hadn't realized you could use top level selectors when running a query selector on a specific element, and doing so should probably help to make sure things are matching correctly.

On a side note, one thing I've tried to avoid with making selectors too too specific is that this could also lead to increased maintenance, if the site changes its structure at all, which could make the replacement stop working for existing icons.  I feel like this is generally more likely than a change that causes the extension to replace icons where you don't intent it to.

But yeah, let me know if these new selectors are more along the lines of what you meant or if there are any other changes you want made to them!

csandman commented:
One last tiny change I added, with the more specific GitHub selectors added previously, the symlink/submodule icon was not being identified properly, so I added one more selector to that list to make it work!

Claudiohbsantos reviewed (approved):
Cool, your notes about selectors in the different cases make sense to me, thanks for looking into these again!

This looks great to me! I'll mark it as approved since I don't expect any more changes to be necessary. I'll hold off on merging for a bit just so I can batch the changes that will affect permissions all together in order to avoid disabling the extension for users repeatedly, but I'll try to get this out this week.
